### PR TITLE
[#5][Backend] As a User, I can sign up and sign in/out with a username and password

### PR DESCRIPTION
**Pull request by malparty**
[#5 [Backend] As a User, I can sign up and sign in/out with a username and password](https://github.com/malparty/google-search-ruby/issues/5)

## What happened

Implementation of Devise, to get Web authentication.

## Insight

- sign up (/users/sign_up -- auto redirected when no logged)
- sign in ( /users/sign_in)
- sign out (button on home page once logged)
- edit profile ( /users/edit)

## Proof Of Work

| image | image |
|------|------|
| image | image |
| image |

### CODE REVIEW

File: spec/support/authentication_helper.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module AuthenticationHelper
+  def sign_in(user = nil)
```
Comment (ankitkalia1195) [inline spec/support/authentication_helper.rb]: @malparty Signing in this way everytime can slow down the test, how about having this way of sign_in only of sign_in tests, and for remaining we just try to set the session. Refer to this project at nimble called mia, https://github.com/nimblehq/mia-web/blob/development/spec/support/devise_helpers.rb.

File: spec/fabricators/user_fabricator.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Fabricator(:user) do
+  lastname 'DENT'
```
Comment (ankitkalia1195) [inline spec/fabricators/user_fabricator.rb]: We can try to use https://github.com/ffaker/ffaker gem as well if required, for generating fake data

**Comment by ankitkalia1195** [comment]
@malparty Why is the doorkeeper required in this PR, as this PR does not relate to api 🤔

**Comment by malparty** [comment]
> @malparty Why is the doorkeeper required in this PR, as this PR does not relate to api 🤔

@ankitkalia1195 I follow up a tutorial and only after already installing Doorkeeper I understood that Doorkeeper was not needed for basic web authentication. So I did not perform the API implementation but I believed that removing Doorkeeper and adding it again later would only add complexity. So I kept it as is.

Now it's much clearer the difference between Devise (basic auth) & DoorKeeper (token auth via oauth2) :)

File: spec/support/authentication_helper.rb (hunk 1 above)
Comment (malparty) [reply spec/support/authentication_helper.rb]: Ok @ankitkalia1195 
What I've changed: 
- Review modules for the helper (more consistent)
- Rename the manual sign_in to "sign_in_ui" (which remain needed to test the login screens)
- Add a "sign_in" helper that utilize provided helpers
- Add a test (when user sign in and reach home page, it should not redirect him to sign in) in order to ensure the "sign_in" (non ui) helper works well

File: spec/fabricators/user_fabricator.rb (hunk 2 above)
Comment (malparty) [reply spec/fabricators/user_fabricator.rb]: I'll try it after lunch! Cool!! :)

Comment (malparty) [reply spec/fabricators/user_fabricator.rb]: Solved! Wow, very easy to use! :)

File: spec/support/authentication_helper.rb (hunk 1 above)
Comment (ankitkalia1195) [reply spec/support/authentication_helper.rb]: @malparty Thanks, have you already committed these changes or yet to commit 🤔

File: spec/systems/login_spec.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe 'login', type: :system do
+  context 'when non logged user reach the app' do
+    it 'redirect the user to the login page' do
```
Comment (ankitkalia1195) [inline spec/systems/login_spec.rb]: small typo, can be `redirects the user` :)

File: db/migrate/20210609041826_add_name_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class AddNameToUsers < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :firstname, :string
```
Comment (ankitkalia1195) [inline db/migrate/20210609041826_add_name_to_users.rb]: If its mandatary field, can we had `null: false`

File: spec/systems/signup_spec.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  context 'when an non logged user sign up with bad password confirmation' do
+    it 'display an error message' do
+      sign_up_ui 'good@email.com', 'complex123password', 'differentPassword123'
+
+      expect(page).to have_selector('#error_explanation li')
+    end
+  end
+
+  context 'when a logged user reach the sign up page' do
+    it 'redirect him to the root_page' do
```
Comment (ankitkalia1195) [inline spec/systems/signup_spec.rb]: it `redirects him to the root_page`

File: spec/systems/signup_spec.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  context 'when an non logged user sign up with bad password confirmation' do
+    it 'display an error message' do
+      sign_up_ui 'good@email.com', 'complex123password', 'differentPassword123'
+
+      expect(page).to have_selector('#error_explanation li')
+    end
+  end
+
+  context 'when a logged user reach the sign up page' do
```
Comment (ankitkalia1195) [inline spec/systems/signup_spec.rb]: `when a logged in user reaches the sign up page`

File: spec/systems/signup_spec.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  context 'when an non logged user sign up with too short password' do
+    it 'display an error message' do
+      sign_up_ui 'good@email.com', '123'
+
+      expect(page).to have_selector('#error_explanation li')
+    end
+  end
+
+  context 'when an non logged user sign up with bad password confirmation' do
```
Comment (ankitkalia1195) [inline spec/systems/signup_spec.rb]: can it be simplified to `when a logged out`

File: spec/systems/signup_spec.rb (hunk 6 above)
Comment (ankitkalia1195) [reply spec/systems/signup_spec.rb]: or how about `when an authenticated user reaches the sign up page` :).

File: spec/systems/signup_spec.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
```
Comment (ankitkalia1195) [inline spec/systems/signup_spec.rb:3]: @malparty How about changing the naming to `authenticated`/ `unauthenticated` everywhere instead of `non logged`/ `logged`. Same for the file above

File: spec/systems/signup_spec.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    it 'displays password confirmation field ' do
+      visit new_user_registration_path
+      expect(find('form')).to have_field('user_password_confirmation')
+    end
+  end
+
+  context 'when non logged user sign up with good data' do
+    it 'create an account and log the user in' do
```
Comment (ankitkalia1195) [inline spec/systems/signup_spec.rb]: creates an account and logs the user in

File: spec/systems/login_spec.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  context 'when valid credentials' do
+    it 'log the user in and display a flash message' do
+      sign_in_ui
+
+      expect(page).to have_content(I18n.t('devise.sessions.signed_in'))
+    end
+  end
+
+  context 'when invalid credentials' do
+    it 'refuse login and display an error' do
```
Comment (ankitkalia1195) [inline spec/systems/login_spec.rb]: `refuses login and displays an error`

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  validates :email, format: URI::MailTo::EMAIL_REGEXP
+
+  # the authenticate method from devise documentation
+  def self.authenticate(email, password)
```
Comment (ankitkalia1195) [inline app/models/user.rb:10]: Why do we need this method, if it provied by default from devise it should be automatically available?

File: app/models/user.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  validates :email, format: URI::MailTo::EMAIL_REGEXP
```
Comment (ankitkalia1195) [inline app/models/user.rb]: Instead of adding this line, can we change the configuration of devise to do this in `config/devise.rb`. Because going this way will add 2 validation, devise already validates email using its own regex.

Comment (malparty) [reply app/models/user.rb]: Instructive!! Thanks

File: app/models/user.rb (hunk 11 above)
Comment (malparty) [reply app/models/user.rb:10]: This is a part for Doorkeeper (on the API task). Not needed for Devise but used in Dookeerper config:
image

Comment (malparty) [reply app/models/user.rb:10]: I followed this tutorial: https://rubyyagi.com/rails-api-authentication-devise-doorkeeper/

Comment (malparty) [reply app/models/user.rb:10]: And the methods is not "in devise code" but it is provided by Devise as a way to do:
https://github.com/heartcombo/devise/wiki/How-To:-Find-a-user-when-you-have-their-credentials

File: db/migrate/20210609041826_add_name_to_users.rb (hunk 4 above)
Comment (malparty) [reply db/migrate/20210609041826_add_name_to_users.rb]: I believe it is not mandatory (at least for the sake of Keywords searching). 
But good to know ;-)

File: spec/support/authentication_helper.rb (hunk 1 above)
Comment (malparty) [reply spec/support/authentication_helper.rb]: already committed/pushed ;-)

File: spec/systems/signup_spec.rb [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+describe 'signup', type: :system do
+  context 'when an unauthenticated user reaches the sign up page' do
+    it 'displays the email field' do
+      visit new_user_registration_path
+      expect(find('form')).to have_field('user_email')
+    end
+
+    it 'displays the password field ' do
+      visit new_user_registration_path
+      expect(find('form')).to have_field('user_password')
```
Comment (ankitkalia1195) [inline spec/systems/signup_spec.rb:16]: Here we can add line break between visit and expect as well.

**Comment by junan** [comment]
@malparty I think  youn can rebase with latest `develop` branch with your local branch so that the those already merged(to `development`) changes will not show this PR and it helps reviewers to review PRs. Currently many new changes, those already exist in the `develop` branch.

File: Gemfile [hunk 14]
```diff
@@ -17,6 +17,7 @@ gem 'i18n-js', '3.5.1' # A library to provide the I18n translations on the Javas
 # Authentications & Authorizations
 gem 'devise' # Authentication solution for Rails with Warden
 gem 'pundit' # Minimal authorization through OO design and pure Ruby classes
+gem 'doorkeeper' # Awesome OAuth 2 provider for your Rails / Grape app
```
Comment (junan) [inline Gemfile:20]: I think this only needed for API, usually we handle all API related changes in a separate PR(ex PR title: [API] As user ...) and  separate task (ex: [API] As user ...). Usully we don't do mix changes API and Web in a single PR. For upcoming PR you can separate changes in a separate PR 👍  (API related changes will be in separate API PR, Web related changes will be in separate Web PR).

File: app/views/devise/shared/_error_messages.html.erb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+<% if resource.errors.any? %>
+  <div id="error_explanation">
+    <h2>
+      <%= I18n.t("errors.messages.not_saved",
```
Comment (junan) [inline app/views/devise/shared/_error_messages.html.erb]: I think  it works without `I18n` part
```suggestion
      <%= t("errors.messages.not_saved",
```

File: db/migrate/20210609041601_devise_create_users.rb [hunk 16]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class DeviseCreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :email,              null: false, default: ""
```
Comment (junan) [inline db/migrate/20210609041601_devise_create_users.rb]: Based on our convention, email field's type should be `citext` 
Screen Shot 2021-06-14 at 2 46 32 PM

File: db/migrate/20210609041826_add_name_to_users.rb (hunk 4 above)
Comment (junan) [inline db/migrate/20210609041826_add_name_to_users.rb]: In Rails, we prefer to use snake case while naming
```suggestion
    add_column :users, :first_name, :string
```

File: db/seeds.rb [hunk 17]
```diff
@@ -5,3 +5,13 @@
...
 #   Character.create(name: 'Luke', movie: movies.first)
+#
+
+# if there is no OAuth application created, create them
+if Doorkeeper::Application.count.zero?
+  Doorkeeper::Application.create(name: "iOS client", redirect_uri: "", scopes: "")
+  Doorkeeper::Application.create(name: "Android client", redirect_uri: "", scopes: "")
+end
+
+
```
Comment (junan) [inline db/seeds.rb]: Lets remove extra space
```suggestion
```

File: spec/fabricators/user_fabricator.rb [hunk 18]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require 'ffaker'
```
Comment (junan) [inline spec/fabricators/user_fabricator.rb]: I think we don't need to require it, it should be available automatically here

File: spec/support/authentication_helper.rb [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+      fill_in 'user_email',    with: user.email
+      fill_in 'user_password', with: user.password
+      click_button 'Log in'
+    end
+
+    def sign_in(user)
+      login_as user, scope: :user
+    end
+
+    def sign_up_ui(email, password, password_confirm = nil)
```
Comment (junan) [inline spec/support/authentication_helper.rb:26]: Lets separate each section with a new line for better redability
```ruby
user = Fabricate(:user)

visit new_user_registration_path

fill_in 'user_firstname', with: user.firstname
fill_in 'user_lastname', with: user.lastname
fill_in 'user_email', with: email
fill_in 'user_password', with: password
fill_in 'user_password_confirmation', with: password_confirm || password

click_button 'Sign up'
```

File: Gemfile (hunk 14 above)
Comment (malparty) [reply Gemfile:20]: Well noted, I'll be more careful for the next ones! ;-) Thanks

File: app/views/devise/shared/_error_messages.html.erb (hunk 15 above)
Comment (malparty) [reply app/views/devise/shared/_error_messages.html.erb]: Do you know why it works in views while it does not work if I just call "t()" in my RSpec tests? :)
Thanks,

Comment (ankitkalia1195) [reply app/views/devise/shared/_error_messages.html.erb]: Because `.t` method is a part of helpers, which are only available in view https://api.rubyonrails.org/v5.1.7/classes/ActionView/Helpers/TranslationHelper.html. You can try the following command by opening up rails console. 

```
   rails c (open up rails console)

   include ActionView::Helpers::TranslationHelper
   t("errors.messages.not_saved")
```

File: db/migrate/20210609041826_add_name_to_users.rb (hunk 4 above)
Comment (malparty) [reply db/migrate/20210609041826_add_name_to_users.rb]: yes, more an English issue from me than a rails convention issue here ˆˆ, 
I've renamed it (db migration added + model refactor).

File: db/migrate/20210614164900_change_email_to_citext_for_users.rb [hunk 20]
```diff
@@ -0,0 +1,6 @@
+class ChangeEmailToCitextForUsers < ActiveRecord::Migration[6.1]
```
Comment (ankitkalia1195) [inline db/migrate/20210614164900_change_email_to_citext_for_users.rb]: @malparty Since the migration to create users is not yet merged on the heroku database, it can be better to change the type in origination migration instead of creating a new one. On your local system, we can rollback the migration and run again

File: db/migrate/20210614165400_rename_names_for_users.rb [hunk 21]
```diff
@@ -0,0 +1,6 @@
+class RenameNamesForUsers < ActiveRecord::Migration[6.1]
+  def change
+    rename_column :users, :lastname, :last_name
```
Comment (ankitkalia1195) [inline db/migrate/20210614165400_rename_names_for_users.rb]: @malparty Same as comment above, changes can be done to same migration instead of creating new migration, as the migration is not yet run on server.

File: spec/fabricators/user_fabricator.rb (hunk 18 above)
Comment (malparty) [reply spec/fabricators/user_fabricator.rb]: When I read comments like this, I feel like "beginner mistake" ! :P 
Thanks!

File: db/migrate/20210614164900_change_email_to_citext_for_users.rb (hunk 20 above)
Comment (malparty) [reply db/migrate/20210614164900_change_email_to_citext_for_users.rb]: All right! I'm on it! (I'll include the rename of last_name first_name too then) :) 
Thanks!

File: db/migrate/20210614165400_rename_names_for_users.rb (hunk 21 above)
Comment (malparty) [reply db/migrate/20210614165400_rename_names_for_users.rb]: Yep, done! :)

File: db/migrate/20210614164900_change_email_to_citext_for_users.rb (hunk 20 above)
Comment (malparty) [reply db/migrate/20210614164900_change_email_to_citext_for_users.rb]: Done! ;-)

File: app/views/devise/shared/_error_messages.html.erb (hunk 15 above)
Comment (malparty) [reply app/views/devise/shared/_error_messages.html.erb]: Thanks! Very clear

**Review by junan — APPROVED** [review]

**Review by ankitkalia1195 — APPROVED** [review]
